### PR TITLE
vis_gen component

### PR DESCRIPTION
**Pull request by airbagy**

### CODE REVIEW

File: utils.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
+# Copyright (C) 2020 and later: Google, Inc.
+
+import os
+import random
+import shutil
+
+def _check_create_dir(dir_path):
```
Comment (FrankYFTang) [inline utils.py]: No need to write this function, just use 
os.makedirs(name, exist_ok=True)
instead
see https://docs.python.org/3/library/os.html

File: utils.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        num_train: Int, total number of training records.
+        num_test: Int, total number of test records.
+
+    Raises:
+        OSError: If no images found in out_dir.
+        ValueError: If num_test is larger than total number of records.
+        OSError: If test data already exists.
+    """
+    # Get absolute path to train and test data directory
+    train_dir_abs = os.path.join(os.getcwd(), train_dir)
```
Comment (FrankYFTang) [inline utils.py]: train_dir_abs = os.path.abspath(train_dir)
see https://docs.python.org/3/library/os.path.html

File: utils.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    test_dir_abs = os.path.join(os.getcwd(), test_dir)
```
Comment (FrankYFTang) [inline utils.py:31]: test_dir_abs = os.path.abspath(test_dir)
see https://docs.python.org/3/library/os.path.html

File: utils.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+    # Create test dir
+    _check_create_dir(test_dir_abs)
```
Comment (FrankYFTang) [inline utils.py]: os.makedirs(test_dir_abs, exist_ok=True)
see https://docs.python.org/3/library/os.html

File: utils.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+    # Get total number of training records
+    num_total = len([name for name in os.listdir(train_dir_abs)
```
Comment (FrankYFTang) [inline utils.py]: num_total  = len(next(os.walk(train_dir_abs))[2])
see https://stackoverflow.com/questions/2632205/how-to-count-the-number-of-files-in-a-directory-using-python

File: utils.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+                     if os.path.isfile(os.path.join(train_dir_abs, name))])
+    num_exist = len([name for name in os.listdir(test_dir_abs)
```
Comment (FrankYFTang) [inline utils.py]: num_exist  = len(next(os.walk(test_dir_abs))[2])
see https://stackoverflow.com/questions/2632205/how-to-count-the-number-of-files-in-a-directory-using-python

File: utils.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    num_train = num_total - num_test
+
+    # Do train/test split
+    print('Creating train test split with {} total records...'
+          .format(num_total))
+    print('Train size: {}'.format(num_train))
+    print('Test size: {}'.format(num_test))
+    filenames = random.sample(os.listdir(train_dir_abs), 100)
+    for filename in filenames:
+        srcpath = os.path.join(train_dir_abs, filename)
```
Comment (FrankYFTang) [inline utils.py:55]: why not just
os.rename(filename, filename, src_dir_fd= train_dir_abs, dst_dir_fd= test_dir_abs)
see https://docs.python.org/3/library/os.html

File: utils.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    if num_exist != 0:
+        raise OSError('Test data already exists.')
+    num_train = num_total - num_test
+
+    # Do train/test split
+    print('Creating train test split with {} total records...'
+          .format(num_total))
+    print('Train size: {}'.format(num_train))
+    print('Test size: {}'.format(num_test))
+    filenames = random.sample(os.listdir(train_dir_abs), 100)
```
Comment (FrankYFTang) [inline utils.py]: filenames = random.sample(next(os.walk(train_dir_abs))[2], num_test)

listdir will give you file and directory.
also you should pass in num_test, not 100 as k to random.sample.

**Review by FrankYFTang — COMMENTED** [review]
my review of utils.py

File: vis_gen.py [hunk 9]
```diff
@@ -99,63 +107,109 @@ def __init__(self, font_size=16, image_size=20,
...
+        """
         return self._grayscale
 
 
     @font_size.setter
     def font_size(self, font_size):
         """Check if font_size is 
             1. larger than 0
-            2. smaller image_size"""
+            2. smaller image_size
```
Comment (FrankYFTang) [inline vis_gen.py]: How about rewrite it as 
Raise ValueError if font_size is 0 or smaller or if font_size is larger than image_size.

File: vis_gen.py [hunk 10]
```diff
@@ -99,63 +107,109 @@ def __init__(self, font_size=16, image_size=20,
...
         else:
             self._font_size = font_size
 
     @image_size.setter
     def image_size(self, image_size):
-        """Check if image_size is larger than 0."""
+        """Check if image_size is:
+            1. larger than 0
+            2. larger than font_size
+            Otherwise, raise ValueError."""
```
Comment (FrankYFTang) [inline vis_gen.py]: Raise ValueError if image_size is smaller than font_size.

**Review by FrankYFTang — COMMENTED** [review]
Great job on newly added tests! Fix the issue in utils.py and similar code in vis_gen.py and you are good to go for this PR.

File: utils.py (hunk 8 above)
Comment (airbagy) [reply utils.py]: Ah that's a huge mistake. Did not realize this since I used 100 as data size only.

File: utils.py (hunk 7 above)
Comment (airbagy) [reply utils.py:55]: I still think the original code have easier to understand logic.

**Review by FrankYFTang — APPROVED** [review]
LGTM @sffc Any comments?

**Review by sffc — COMMENTED** [review]
The code is well-documented and Pythonic!  I didn't think through every code path or edge case, but based on what I saw, it looks like it will be very easy to understand this code when needed.

Optional: instead of `DATA_FOLDER`, it's common to create a file named `.gitkeep`.

https://stackoverflow.com/questions/7229885/what-are-the-differences-between-gitignore-and-gitkeep

**Review by sffc — APPROVED** [review]

**Pull request merged**